src/java.base/share/classes/jdk/crac/Core.java

```diff
@@ -206,4 +227,20 @@ private static void checkpointRestoreInternal() {
         thread.setDaemon(true);
         thread.start();
     }
+
+    static void addLock(CheckpointLock lock) {
+        synchronized(criticalSectionsLock) {
+            synchronized (lockQ) {
+                lockQ.add(lock);
+            }
+        }
+    }
+
+    static void removeLock(CheckpointLock lock) {
+        synchronized (lockQ) {
+            lockQ.remove(lock);
+            lockQ.notify();
```

Comment:
This will wake up the `checkpointRestore1` wait loop on every lock removal.  If there are a lot of CheckpointLock's this code will spend a lot of time waking up the other loop only for it to check and go back to sleep.

Would it make more sense to only `notify` the `lockQ` when the queue is empty?

Reply:
Good catch. Thank you. 
Even if notify does not really wakeup another thread, it would be better to notify on empty queue only